# Statistics
- random variable
- basic statistics (numeric)
  - min, max
  - mean, std, median
  - 25- and 75-percentiles
  - distribution
- basic statistics (categoric)
- t-testing
- bootstrap

### Prepare random variable to demonstrate the results

In [1]:
# prepare random vector to demonstrate the results

import random

random.seed(42)

# make random vector
size = 1500
xs = [0] * size  # make array with size elements and fill it with 0s
for k in range(size):
    xs[k] = random.random()
    
xs[:5], xs[-5:]

([0.6394267984578837,
  0.025010755222666936,
  0.27502931836911926,
  0.22321073814882275,
  0.7364712141640124],
 [0.9479043786030863,
  0.07504116498815927,
  0.6375125378832983,
  0.3633111306509823,
  0.8010959755621699])

In [2]:
# ex 1.3.X: make random vector with values uniformly distributed from -1 to 1
# ex 1.3.X: make random matrix, size 30x120, with values uniformly distributed from -1 to 3

In [3]:
# ex

random.seed(42)

size = 100
xs = [0] * size
for k in range(size):
    xs[k] = 1 - 2 * random.random()
    
xs[:5], xs[-5:], min(xs), max(xs)

([-0.2788535969157675,
  0.9499784895546661,
  0.4499413632617615,
  0.5535785237023545,
  -0.4729424283280248],
 [0.23676142698692648,
  -0.9922427604801936,
  -0.058228690198274036,
  -0.9421567552272363,
  -0.7215594044689961],
 -0.9950752129902205,
 0.987002480643878)

In [4]:
# ex

random.seed(42)

r, c = 30, 120
m = []
for row in range(r):
    cur_row = [0] * c
    for col in range(c):
        cur_row[col] = 4 * random.random() - 1
    m.append(cur_row)
    
max(m[29]), min(m[29]), m[29][119]

(2.9511664990459088, -0.9993211380866542, 2.742048309850143)

### Basic statistics (numeric variable)
- min, max
- mean, sd, median
- 25- and 75-percentiles
- distribution

In [5]:
# min, max

min_x = min(xs)
max_x = max(xs)
min_x, max_x

(-0.9950752129902205, 0.987002480643878)

In [6]:
# mean

mean_x = 0
for x in xs:
    mean_x += x

mean_x /= len(xs)
mean_x

0.04086151374716617

In [7]:
# std

import math  # package with math functions

std_x = 0
for x in xs:
    std_x += (x - mean_x) ** 2
    
std_x = math.sqrt(std_x / (len(xs) - 1))
std_x

0.5900078691486795

In [8]:
# median

median = sorted(xs)[int(size/2)]
median

0.0822962948252024

In [9]:
# percentiles

sorted_xs = sorted(xs)  # presort the array to avoid repetitive computations
quart1 = sorted_xs[int(size/4)]
quart3 = sorted_xs[int(3 * size/4)]
quart1, quart3

(-0.40914367242984695, 0.5535785237023545)

In [10]:
# simple histogram

bin_count = 12
# bin_size = 1 / bin_count  # because xs was generated as random variable from [0, 1]
bin_size = (max_x - min_x) / bin_count  # more general way
hist = [0] * bin_count
for x in xs:
    bin_idx = int(x / bin_size)
    hist[min(bin_idx, bin_count - 1)] += 1
print("histogram: ", hist)  # should be ~size / bin_count for each bin
print("'ideal' size of each bin: ", size / bin_count)  # note usage of quotes inside quotesa

histogram:  [12, 9, 9, 12, 7, 10, 2, 7, 7, 6, 7, 12]
'ideal' size of each bin:  8.333333333333334


### Basic statistics (categorical variable)
- unique elements
- count

In [11]:
random.seed(42)

cats = [""] * size
for k in range(size):
    cats[k] = random.choice(["a", "b", "c", "d"])
cats[:5], cats[-5:]

(['a', 'a', 'c', 'b', 'b'], ['b', 'b', 'd', 'a', 'd'])

In [12]:
# count unique elements
unique_cats = set(cats)
unique_cats

{'a', 'b', 'c', 'd'}

In [13]:
# ex 1.3.X: implement detection of unique categories, without using set, but by using a loop

In [14]:
# ex

unique_cats_count = dict()
for cat in unique_cats:
    unique_cats_count[cat] = 0
    
for cat in cats:
    unique_cats_count[cat] += 1
    
unique_cats_count

{'b': 29, 'c': 24, 'd': 20, 'a': 27}

### Statistical hypothesis testing (on the example of t-test)
[`t-test`](https://en.wikipedia.org/wiki/Student%27s_t-test) is one of the most commonly used statistical tool to check whether two distribution are different or not.  
`t-test` is applicable when data is distributed **normally**. If the condition is violated, `t-test` can give meaningless results.

In [15]:
random.seed(42)

# dist1
size1 = 200
xs1 = [0] * size1
mean1 = 0

for k in range(size1):
    xs1[k] = random.gauss(20, 2)
    mean1 += xs1[k]
mean1 /= size1

# dist2
size2 = 150
xs2 = [0] * size2
mean2 = 0

for k in range(size2):
    xs2[k] = random.gauss(21, 25)
    mean2 += xs2[k]
mean2 /= size2

mean1, mean2

(20.232259256634066, 22.07622069136082)

Imagine that we only know information about means.  
**`20.23` and `22.07` can be considered as similar?**, Does this mean that distributions are similar as well? Let's find out!

In [16]:
# calcuate variances

var1 = 0
for x in xs1:
    var1 += (x - mean1) ** 2
var1 /= size1 - 1

var2 = 0
for x in xs2:
    var2 += (x - mean2) ** 2
var2 /= size2 - 1

var1, var2

(3.416010615296955, 807.0836906738768)

**They are different!** What does `t-test` say?  
We will be using `t-test` samples with unequal variances (formulas are from [**Welch's t-test**](https://en.wikipedia.org/wiki/Welch%27s_t-test)

$${\displaystyle t={\frac {\Delta {\overline {X}}}{s_{\Delta {\bar {X}}}}}={\frac {{\overline {X}}_{1}-{\overline {X}}_{2}}{\sqrt {{s_{{\bar {X}}_{1}}^{2}}+{s_{{\bar {X}}_{2}}^{2}}}}}\,}$$  
$${\displaystyle s_{{\bar {X}}_{i}}={s_{i} \over {\sqrt {N_{i}}}}\,}$$

In [17]:
s_delta = math.sqrt(var1 / size1 + var2 / size2)
t_statistics = (mean1 - mean2) / s_delta
t_95_critical_value = 1.960 # from https://en.wikipedia.org/wiki/Student%27s_t-distribution#Table_of_selected_values

if abs(t_statistics) < t_95_critical_value:
    print("Distribution are similar ({})".format(t_statistics))
    # it means that p-val is greater than 0.05, (accept the hypothesis)
    # which menas null hypothesis (xs1 & xs2 are equal) is true
else:
    print("Distribution are different ({})".format(t_statistics))
    # it means that p-val is lower than 0.05, (reject the hypothesis)
    # which means null hypothesis (x1 & x2 are equal) is not true

Distribution are similar (-0.7936882608550769)


In [18]:
# ex 1.3.X: Find parameters for the xs2, not equal to thos for xs1, such that distributions become different

In [19]:
# ex

def generate_random_sample(mu, std, size=150, seed=42):
    random.seed(seed)
    
    xs = [0] * size
    mean = 0
    for k in range(size):
        xs[k] = random.gauss(mu, std)
        mean += xs[k]
    mean /= size
    
    var = 0
    for x in xs:
        var += (x - mean) ** 2
    var /= size - 1
    
    return (mean, var, size)

def calculate_t(control_param, experiment_param):
    mean1, var1, size1 = control_param
    mean2, var2, size2 = experiment_param
    
    s_delta = math.sqrt(var1 / size1 + var2 / size2)
    return (mean1 - mean2) / s_delta

def find_diff_param(mus, stds, control_param, critical_value=1.960):
    result = []
    for mu in mus:
        for std in stds:
            experiment_param = generate_random_sample(mu, std)
            t_statistic = calculate_t(control_param, experiment_param)
            if abs(t_statistic) > critical_value:
                param = {
                    "mu": mu,
                    "std": std,
                    "t-stats": t_statistic,
                }
                result.append(param)
    
    return result


control_param = (mean1, var1, size1)
mus = range(18, 22)
stds = range(0, 10)

print(np.mean(xs1), np.std(xs1))

result = find_diff_param(mus, stds, control_param)
result

20.232259256634066 1.8436188766175263


[{'mu': 18, 'std': 0, 't-stats': 17.080478167633142},
 {'mu': 18, 'std': 1, 't-stats': 14.233819718983984},
 {'mu': 18, 'std': 2, 't-stats': 10.332241052563173},
 {'mu': 18, 'std': 3, 't-stats': 7.578485463229778},
 {'mu': 18, 'std': 4, 't-stats': 5.780284101261792},
 {'mu': 18, 'std': 5, 't-stats': 4.564148761253749},
 {'mu': 18, 'std': 6, 't-stats': 3.701011649616832},
 {'mu': 18, 'std': 7, 't-stats': 3.061613426721033},
 {'mu': 18, 'std': 8, 't-stats': 2.570959606034226},
 {'mu': 18, 'std': 9, 't-stats': 2.1834771969024427},
 {'mu': 19, 'std': 0, 't-stats': 9.428822959183874},
 {'mu': 19, 'std': 1, 't-stats': 7.624880046905812},
 {'mu': 19, 'std': 2, 't-stats': 5.35329175498358},
 {'mu': 19, 'std': 3, 't-stats': 3.7828811435700653},
 {'mu': 19, 'std': 4, 't-stats': 2.7667542859098653},
 {'mu': 19, 'std': 5, 't-stats': 2.083058289263658},
 {'mu': 21, 'std': 0, 't-stats': -5.874487457714662},
 {'mu': 21, 'std': 1, 't-stats': -5.592999297250556},
 {'mu': 21, 'std': 2, 't-stats': -4.604

For the cases when at least one distribution is not normal **non-parametric** test should be used

If there is time:
- Add simple test for normality (by analysing 1-, 2-, 3- sigma, IQR, or something like that https://en.wikipedia.org/wiki/Normality_test#Back-of-the-envelope_test)

In [20]:
np.mean(xs1), np.std(xs1)

(20.232259256634066, 1.8436188766175263)

### Bootstrapping

In [21]:
random.seed(42)

size = 1000
xs = [0] * size
mean_x = 0
for k in range(size):
    xs[k] = random.random()
    mean_x += xs[k]
mean_x /= size
    
xs[:5], xs[-5:], min(xs), max(xs)

([0.6394267984578837,
  0.025010755222666936,
  0.27502931836911926,
  0.22321073814882275,
  0.7364712141640124],
 [0.032716286855046905,
  0.3705299587344354,
  0.44338308606070165,
  0.950555169851427,
  0.8554501933059546],
 0.0004059396972875273,
 0.9999078285092092)

In [22]:
# bootstrapping (https://en.wikipedia.org/wiki/Bootstrapping_(statistics))
print("sample mean:", mean_x)
print("true mean:", 0.5)
random.seed(42)
trials = 100
bs_mean_x = 0

for _ in range(trials):
    cur_mean_x = 0
    for _ in xs:
        idx = random.randint(0, size-1)
        cur_mean_x += xs[idx]
    cur_mean_x /= len(xs)
    bs_mean_x += cur_mean_x
bs_mean_x /= trials

bs_mean_x, mean_x # true mean is 0.5
print("bootstrapped mean:", bs_mean_x)

sample mean: 0.5125619702436156
true mean: 0.5
bootstrapped mean: 0.5133657082871199


In [23]:
# ex 1.3.X: check how number of trials affects accuracy of the bootstrapped mean
# ex 1.3.X: write function to compute bootstrapped standard variance (true value is 1/12 ~ 0.083)

### Summary

Computing basic statistics is a good exercise for programming in python (and a simple statistics reminder). Later on we will see how to do it in mocre compact way and how to use packages for that.

List of used packages:
- random
- math